### PR TITLE
Run simplifier in topological order.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
This PR makes the simplifier go through each value definition in topological order, instead of running all at once. This way optimizations can build on previous optimizations.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 1]
```diff
@@ -356,6 +361,31 @@ simplifyExpr world = fst . cata go
...
+        sccs = G.stronglyConnComp . map dvalNode . NM.toList $ moduleValues m
+    in concatMap toList sccs
+
+simplifyDefValue :: World -> DefValue -> DefValue
+simplifyDefValue world dval = dval { dvalBody = simplifyExpr world (dvalBody dval) }
+
 simplifyModule :: World -> Module -> Module
 simplifyModule world m =
-    over moduleExpr (simplifyExpr (extendWorldSelf m world)) m
+    let step !accum dval =
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:386]: isn’t the bang pattern redundant if you use a `foldl'`?

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 2]
```diff
@@ -356,6 +361,31 @@ simplifyExpr world = fst . cata go
...
+            let m' = m { moduleValues = accum }
+                w' = extendWorldSelf m' world
+                d' = simplifyDefValue w' dval
+            in NM.insert d' accum
+    in m { moduleValues = foldl' step NM.empty (topoSortDefValues m) }
```
Comment (cocreature) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:391]: This looks like you are reordering the values. I guess this doesn’t matter for anything?

Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:391]: I _think_ it's ok to reorder.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs (hunk 1 above)
Comment (sofiafaro-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:386]: Right, thanks!

**Pull request merged**